tests/test-find-remotes.sh

```diff
@@ -0,0 +1,225 @@
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+# Boston, MA 02111-1307, USA.
+
+set -euo pipefail
+
+. $(dirname $0)/libtest.sh
+
+echo '1..1'
+
+# Create two upstream collection repositories with some example commits
+cd ${test_tmpdir}
+
+mkdir apps-collection
+ostree_repo_init apps-collection --collection-id apps-collection
+mkdir -p files
+pushd files
+${CMD_PREFIX} ostree --repo=../apps-collection commit -s "Test apps-collection commit 1" -b app1 --gpg-homedir=${TEST_GPG_KEYHOME} --gpg-sign=${TEST_GPG_KEYID_1} > ../app1-checksum
+${CMD_PREFIX} ostree --repo=../apps-collection commit -s "Test apps-collection commit 2" -b app2 --gpg-homedir=${TEST_GPG_KEYHOME} --gpg-sign=${TEST_GPG_KEYID_1} > ../app2-checksum
+popd
+${CMD_PREFIX} ostree --repo=apps-collection summary --update --gpg-homedir=${TEST_GPG_KEYHOME} --gpg-sign=${TEST_GPG_KEYID_1}
+
+mkdir os-collection
+ostree_repo_init os-collection --collection-id os-collection
+mkdir -p files
+pushd files
+${CMD_PREFIX} ostree --repo=../os-collection commit -s "Test os-collection commit 1" -b os/amd64/master --gpg-homedir=${TEST_GPG_KEYHOME} --gpg-sign=${TEST_GPG_KEYID_2} > ../os-checksum
+popd
+${CMD_PREFIX} ostree --repo=os-collection summary --update --gpg-homedir=${TEST_GPG_KEYHOME} --gpg-sign=${TEST_GPG_KEYID_2}
+
+# Create a local repository where we pull the branches from the two remotes as normal, using GPG.
+mkdir local
+ostree_repo_init local
+${CMD_PREFIX} ostree --repo=local remote add apps-remote file://$(pwd)/apps-collection --collection-id apps-collection --gpg-import=${test_tmpdir}/gpghome/key1.asc
+${CMD_PREFIX} ostree --repo=local remote add os-remote file://$(pwd)/os-collection --collection-id os-collection --gpg-import=${test_tmpdir}/gpghome/key2.asc
+
+${CMD_PREFIX} ostree --repo=local pull apps-remote app1
+${CMD_PREFIX} ostree --repo=local pull os-remote os/amd64/master
+
+${CMD_PREFIX} ostree --repo=local refs > refs
+assert_file_has_content refs "^apps-remote:app1$"
+assert_file_has_content refs "^os-remote:os/amd64/master$"
+
+${CMD_PREFIX} ostree --repo=local refs --collections | wc -l > refscount
+assert_file_has_content refscount "^0$"
+
+# Create a local mirror repository where we pull the branches *in mirror mode* from the two remotes.
+# This should pull them into refs/mirrors, since the remotes advertise a collection ID.
```

Comment:
If we have a setup where the upstream repository exposes two collection IDs (C1, C2) each with two refs (R1, R2), it could advertise them in its `summary` file in two different ways.

Case A: refs list empty, collection ID is empty, collection map contains `{ C1: [R1, R2], C2: [R1, R2] }`.

Case B: refs list contains [R1, R2], collection ID is C1, collection map contains `{ C2: [R1, R2] }`.

The difference between these cases is a result of how the owner of the repository sets it up (libostree and the command line tools support both configurations).

If the libostree which is pulling from this upstream repository (to create a mirror, or as a normal client) is old, case A will fail, and case B will result in `refs/heads/R1` and `refs/heads/R2` (which are the C1 refs).

If the libostree which is pulling is new, both cases will result in `refs/mirrors/{C1,C2}/{R1,R2}`. We can’t mirror it to create a repository identical to case B, since the collection ID in the summary is meant to be used for upstream repositories only.

So I think this is a configuration thing. Generally, people deploying repositories should configure them as for case B, with the refs the repository lists in `refs/heads` being those it’s always listed; and the refs in `refs/mirrors` being new, additional ones.

The case we can’t support is an old libostree client pulling from a mirror created with a new libostree (i.e. the `refs/mirrors/{C1,C2}/{R1,R2}` mirror). I don’t think this is a problem in the setups I’ve been thinking about — mirroring would be used over P2P, and in order to access repositories over P2P, the client would have to be using a new libostree anyway.

Does that cover what you had in mind?

> Although I guess this requires using `remote add --collection-id`...so if I want to run a "transparent mirror", I'd omit the collection ID?

If `--collection-id` is omitted from `remote add` when setting up the mirror, I don’t think it’ll make a difference to the `pull --mirror`, since the code in `ostree-repo-pull.c` to process the refs bases it all off the downloaded `summary` file, rather than the remote’s configuration. That `--collection-id` argument is there to allow `find-remotes` to use that remote configuration for P2P resolution of refs in the `find-remotes apps-collection app1` invocation (and others) lower down in the test.